horace_core/sqw/@d2d/private/rebin_2d_1axis.m

```diff
@@ -0,0 +1,236 @@
+end
+if (min(min(xin)))<min(min(xout)-small)
+    error('Rebinning error: minimum of output bin boundaries must be <= min of input bin boundaries');
+end
+
+%==========================================================================
+
+%Before we do anything, we need to find points where npix is zero, or where
+%the errorbar is zero, and set s=e=n=0:
+nin_test=double(nin>0);%matrix of zeros and ones
+sin=sin.*nin_test; ein=ein.*nin_test;
+okerr=double(ein>0);
+sin=sin.*okerr; nin=nin.*okerr;
+
+%Likewise, we must find points where either sin or ein are NaN, and replace
+%with zeros:
+ein(isnan(sin))=0; nin(isnan(sin))=0; sin(isnan(sin))=0;
+sin(isnan(ein))=0; nin(isnan(ein))=0; ein(isnan(ein))=0;
+
+%We also must convert error to fractional error.
+ein_original=ein;
+ein=ein./sin;%fractional variance
+ein(isnan(ein) | isinf(ein))=0;
+
+%Sort according to ascending xin:
+xinlo=xin([1:end-1],[1:end-1]); xinhi=xin([2:end],[1:end-1]);
+%nin=double(nin>0);%make this just have zeros or ones
+
+xtosort=xinlo(:,1);
+
+snew=sortrows([xtosort sin],1); snew(:,1)=[];%2nd command gets rid of 1st col, which specifies x co-ord
+enew=sortrows([xtosort ein],1); enew(:,1)=[];
+nnew=sortrows([xtosort nin],1); nnew(:,1)=[];
+
+%Check for repeated values. This screws things up totally at the moment. A
+%later version of this code will address this point, but throw up an error
+%for now:
+xshift_up=circshift(xin(:,1),-1); xshift_down=circshift(xin(:,1),1);
+if any(xshift_up==xin(:,1)) || any(xshift_down==xin(:,1))
+    error('Rebinning error: input dataset should not have repeated bin boundaries');
+end
+
+%define a small number (to be used later)
+dx=diff(xin(:,1));
+eps=1e-5*min(dx);%ensures that our small number is much smaller that the width of the input bins
+
+%Make vector of lower bin boundaries, and vector of upper bin boundaries
+%for output:
+xout=sort(xout);%ensures that output bin boundaries are also sorted in ascending order
+xlo=xout(1:end-1);
+xhi=xout(2:end);
+
+%Note the largest errorbar in the input dataset:
+maxerr=max(max(ein));
+
+%Intialise the outputs:
+[sz1,sz2]=size(sin);
+sout=[]; eout=sout; nout=sout;
+xinnew=xin(:,1);
+lonew=xinnew(1:end-1); hinew=xinnew(2:end);
+filler=zeros(1,sz2);
+
+%Make xnew and ynew matrices:
+[xnew,ynew]=ndgrid(xout,yin(1,:));
+
+%Now do the rebinning:
+for i=1:length(xlo)
+   cumulsum=[]; esum=[];
+   inbin=(xinnew>xlo(i) & xinnew<xhi(i));%gives us index of a data bin that is inside an output bin
+   ind=find(inbin);%returns the indices of these points
+   if length(ind)==1
+%        %we have a data boundary in output bin, but it is not fully
+%        %contained therein
+       binfrac=[];
+       [cumulsum,esum,nsum]=rebin_2d_1axis_multibins(ind,binfrac,snew,enew,nnew,xinnew,eps,xlo,xhi,i);
+       %
+       %Create final signal, error, npix arrays.
+       if ~isempty(cumulsum)
+           sout=[sout; cumulsum];
+           eout=[eout; esum];
+%            ntmp=ones(size(cumulsum));
```

Comment:
There is dead code is scattered through this function; if it's not needed is there any point in carrying over to the new location?

Reply:
Removed